apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PriorityList.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// A simple sorted list that allows items to return equality for IComparable.CompareTo
+    /// </summary>
+    internal class PriorityList<T> where T : IPriorityListItem<T>
+    {
+        private readonly List<T> _innerList = new List<T>();
+
+        public int Count
+        {
+            get
+            {
+                lock (_innerList)
+                {
+                    return _innerList.Count;
+                }
+            }
+        }
+
+        public IEnumerable<T> GetSnapshot()
+        {
+            lock (_innerList)
+            {
+                return _innerList.ToArray();
+            }
+        }
+
+        public void Add(T item)
+        {
+            lock (_innerList)
+            {
+                var index = _innerList.BinarySearch(item);
+                if (index < 0)
+                {
+                    //if priority changes during this call, it will be blocked from
+                    //processing until after Add has completed.
+                    item.PriorityChanged += Item_PriorityChanged;
+
+                    _innerList.Insert(~index, item);
+                }
+            }
+        }
+
+        public void Remove(T item)
```

Comment:
Yeah I think I discussed this in another comment.  There is another possible design here (of which you give a variation) where either:

a) there is some prioritykey whose state is tracked by the prioritylist and therefore can store earlier versions of said "key"
b) the item itself in some way tracks previous states like by firing a changing event before modification -- which is similar to what the previous code did by removing before modifying state that could change the priority.

I consider "a" the stronger candidate because its less fragile than "b" which can easily break if there is a missing or out of order notification.  "a" is better because the list itself knows with what key an item is currently registered with -- so better encapsulates the functionality within the prioritylist.

At the end I decided that since the complexity is linear anyway for insert, until I do a proper heap there is little value in going through the work.  The lists should be fairly small (# grpc channels probably ~10) as well so putting a lot of work here wont result in any measurable gain.